controller/api/destination/endpoint_translator.go

```diff
@@ -481,43 +481,39 @@ func createWeightedAddr(address watcher.Address, opaquePorts map[uint32]struct{}
 		return &weightedAddr, nil
 	}
 
-	skippedInboundPorts, err := getPodSkippedInboundPortsAnnotations(address.Pod)
-	if err != nil {
-		log.Errorf("failed to get ignored inbound ports annotation for pod: %s", err)
-	}
+	skippedInboundPorts := getPodSkippedInboundPortsAnnotations(address.Pod)
 
 	controllerNSLabel := address.Pod.Labels[pkgK8s.ControllerNSLabel]
 	sa, ns := pkgK8s.GetServiceAccountAndNS(address.Pod)
 	weightedAddr.MetricLabels = pkgK8s.GetPodLabels(address.OwnerKind, address.OwnerName, address.Pod)
 	_, isSkippedInboundPort := skippedInboundPorts[address.Port]
 
-	// If the pod is controlled by any Linkerd control plane, then it can be
-	// hinted that this destination knows H2 (and handles our orig-proto
-	// translation)
 	weightedAddr.ProtocolHint = &pb.ProtocolHint{}
-	if controllerNSLabel != "" && !isSkippedInboundPort {
+
+	_, opaquePort := opaquePorts[address.Port]
+	if address.OpaqueProtocol || opaquePort {
 		// If address is set as opaque by a Server, or its port is set as
-		// opaque by annotation or default value, then hint its proxy's
-		// inbound port, and set the hinted protocol to Opaque, so that the
-		// client proxy does not send a SessionProtocol.
-		_, opaquePort := opaquePorts[address.Port]
-		if address.OpaqueProtocol || opaquePort {
+		// opaque by annotation or default value, then set the hinted protocol to
+		// Opaque.
+		weightedAddr.ProtocolHint.Protocol = &pb.ProtocolHint_Opaque_{
+			Opaque: &pb.ProtocolHint_Opaque{},
+		}
+		if controllerNSLabel != "" && !isSkippedInboundPort {
```

Comment:
Sorry. I'm going to contradict something I said earlier.

Retracing through https://github.com/linkerd/linkerd2/issues/9888, it seems like contract we were establishing for ProtocolHint::Opaque is that protocol hints are only set when the peer is a proxy. To maintain that contract, this needs to be:

```suggestion
		// opaque by annotation or default value, then set the hinted protocol to
		// Opaque.
		if controllerNSLabel != "" && !isSkippedInboundPort {
			weightedAddr.ProtocolHint.Protocol = &pb.ProtocolHint_Opaque_{
				Opaque: &pb.ProtocolHint_Opaque{},
			}
```

Note that this change will likely cause a test to fail.